src/executor/sort_agg.rs

```diff
@@ -0,0 +1,189 @@
+// Copyright 2022 RisingLight Project Authors. Licensed under Apache-2.0.
+#![allow(dead_code)]
+use itertools::Itertools;
+use smallvec::SmallVec;
+
+use super::*;
+use crate::array::{ArrayBuilderImpl, ArrayImpl};
+use crate::binder::BoundAggCall;
+use crate::types::{DataTypeExt, DataTypeKind};
+
+pub struct SortAggExecutor {
+    pub agg_calls: Vec<BoundAggCall>,
+    pub group_keys: Vec<BoundExpr>,
+    pub child: BoxedExecutor,
+}
+
+impl SortAggExecutor {
+    #[try_stream(boxed, ok = DataChunk, error = ExecutorError)]
+    pub async fn execute(self) {
+        let mut states_vec: Vec<SmallVec<[Box<dyn AggregationState>; 16]>> = vec![];
+        let mut last_key = Arc::new(HashKey::new());
+
+        #[for_await]
+        for chunk in self.child {
+            // Eval group keys and arguments
+            let chunk = chunk?;
+            let exprs: SmallVec<[ArrayImpl; 16]> = self
+                .agg_calls
+                .iter()
+                .map(|agg| agg.args[0].eval(&chunk))
+                .try_collect()?;
+            let group_cols: SmallVec<[ArrayImpl; 16]> = self
+                .group_keys
+                .iter()
+                .map(|e| e.eval(&chunk))
+                .try_collect()?;
+
+            let num_rows = chunk.cardinality();
+            for row_idx in 0..num_rows {
+                // Create group key
+                let mut group_key = HashKey::new();
+                for col in group_cols.iter() {
+                    group_key.push(col.get(row_idx));
+                }
+                let group_key = Arc::new(group_key);
+                if last_key.len() != 0 && group_key.eq(&last_key) {
+                    let states = states_vec.last_mut().unwrap();
+
+                    for (state, expr) in states.iter_mut().zip_eq(&exprs) {
+                        state.update_single(&expr.get(row_idx))?;
+                    }
+                } else {
+                    let mut states = create_agg_states(&self.agg_calls);
+                    for (state, expr) in states.iter_mut().zip_eq(&exprs) {
+                        state.update_single(&expr.get(row_idx))?;
+                    }
+                    states_vec.push(states);
```

Comment:
Why not yield the agg result of last group here? If so, there's no need to store all previous states.

Reply:
When you have time, hope to help review again. I'm not familiar with async in rust, if there are mistakes, please help point them out (by the way, there are related materials that can also be recommended.  (：thx

Reply:
Be confident!

Reply:
BTW, you don't need to be familiar with async actually. You can just "feel it". 😄

Specifically, to finish this work you just need to know the **semantics** of `yield` and generators.